pkg/code-generator/schemagen/generator.go

```diff
@@ -0,0 +1,125 @@
+package schemagen
+
+import (
+	"fmt"
+
+	"github.com/solo-io/solo-kit/pkg/code-generator/collector"
+
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/solo-io/go-utils/log"
+	"github.com/solo-io/solo-kit/pkg/code-generator/model"
+)
+
+type ValidationSchemaOptions struct {
+	// Path to the directory where CRDs will be read from and written to
+	CrdDirectory string
+
+	// Tool used to generate JsonSchemas, defaults to protoc
+	JsonSchemaTool string
+}
+
+type JsonSchemaGenerator interface {
+	GetJsonSchemaForProject(project *model.Project) (map[schema.GroupVersionKind]*v1beta1.JSONSchemaProps, error)
+}
+
+func GenerateOpenApiValidationSchemas(project *model.Project, options *ValidationSchemaOptions, importsCollector collector.Collector, absoluteRoot string) error {
+	if options == nil || options.CrdDirectory == "" {
+		log.Debugf("No CRDDirectory provided, skipping schema-gen")
+		return nil
+	}
+
+	if !project.ProjectConfig.GenKubeValidationSchemas {
+		log.Debugf("Project %s not configured to generate validation schema", project.String())
+		return nil
+	}
+
+	// Extract the CRDs from the directory
+	crds, err := GetCRDsFromDirectory(options.CrdDirectory)
+	if err != nil {
+		return err
+	}
+
+	if len(crds) == 0 {
+		log.Debugf("Found 0 CRDs in directory: %s, skipping schema-gen", options.CrdDirectory)
+		return nil
+	}
+
+	// Build the JsonSchemas for the project
+	var jsonSchemaGenerator JsonSchemaGenerator
+	switch options.JsonSchemaTool {
+	case "cue":
+		jsonSchemaGenerator = NewCueGenerator(importsCollector, absoluteRoot)
+	case "protoc":
+		jsonSchemaGenerator = NewProtocGenerator(importsCollector, absoluteRoot)
+	default:
+		jsonSchemaGenerator = NewProtocGenerator(importsCollector, absoluteRoot)
+	}
+
+	jsonSchemasByGVK, err := jsonSchemaGenerator.GetJsonSchemaForProject(project)
+	if err != nil {
+		return err
+	}
+
+	// For each matching CRD, apply the JSON schema to that CRD
+	// Use Group.Version.Kind to match CRDs and Schemas
+	crdWriter := NewCrdWriter(options.CrdDirectory)
+	for _, crd := range crds {
+		crdGVK := schema.GroupVersionKind{
+			Group:   crd.Spec.Group,
+			Version: crd.Spec.Version,
+			Kind:    crd.Spec.Names.Kind,
+		}
+
+		specJsonSchema, ok := jsonSchemasByGVK[crdGVK]
+		if !ok {
+			continue
+		}
+
+		removeProtoMetadataValidation(specJsonSchema)
```

Comment:
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema ... `if metadata is specified, then only restrictions on metadata.name and metadata.generateName are allowed.`

The kube api server is responsible for managing this field, so users are not allowed to define schemas on it. We remove validation altogether.

Reply:
add this comment to code?